tests/unit/models/gpflow/test_models.py

```diff
@@ -939,3 +906,104 @@ def test_gaussian_process_regression_conditional_predict_f_sample() -> None:
         sample_cov = tfp.stats.covariance(samples[i, :, :, 0], sample_axis=0)
         np.testing.assert_allclose(sample_mean, predj_meani, atol=1e-2, rtol=1e-2)
         np.testing.assert_allclose(sample_cov, predj_covi[0], atol=1e-2, rtol=1e-2)
+
+
+@pytest.mark.parametrize("num_outputs", [1, 2])
+def test_gaussian_process_regression_pairwise_covariance(num_outputs: int) -> None:
+    x = tf.constant(np.arange(1, 5).reshape(-1, 1), dtype=gpflow.default_float())  # shape: [4, 1]
+    y = fnc_3x_plus_10(x)
+    model = GaussianProcessRegression(gpr_model(x, tf.repeat(y, num_outputs, axis=1)))
+
+    query_points_1 = tf.concat([0.5 * x, 0.5 * x], 0)  # shape: [8, 1]
+    query_points_2 = tf.concat([2 * x, 2 * x, 2 * x], 0)  # shape: [12, 1]
+
+    all_query_points = tf.concat([query_points_1, query_points_2], 0)
+    _, predictive_covariance = model.predict_joint(all_query_points)
+    expected_covariance = predictive_covariance[:, :8, 8:]
+
+    actual_covariance = model.covariance_between_points(query_points_1, query_points_2)
+
+    np.testing.assert_allclose(expected_covariance, actual_covariance, atol=1e-5)
+
+
+@random_seed
+def test_gpflow_models_pairwise_covariance(gpflow_interface_factory: ModelFactoryType) -> None:
+    x = tf.constant(np.arange(1, 5).reshape(-1, 1), dtype=gpflow.default_float())  # shape: [4, 1]
+    y = fnc_3x_plus_10(x)
+    model, _ = gpflow_interface_factory(x, y)
+
+    if isinstance(model.model, SGPR):
+        pytest.skip("Covariance between points is not supported for SGPR.")
+
+    if isinstance(model.model, GPR):
+        model.optimize(Dataset(x, y))
+    elif isinstance(model.model, (VGP, SVGP)):  # for speed just update q_sqrt rather than optimize
+        num_inducing_points = tf.shape(model.model.q_sqrt)[1]
+        sampled_q_sqrt = tfp.distributions.WishartTriL(5, tf.eye(num_inducing_points)).sample(1)
+        model.model.q_sqrt.assign(sampled_q_sqrt)
+
+    query_points_1 = tf.concat([0.5 * x, 0.5 * x], 0)  # shape: [8, 1]
+    query_points_2 = tf.concat([2 * x, 2 * x, 2 * x], 0)  # shape: [12, 1]
+
+    all_query_points = tf.concat([query_points_1, query_points_2], 0)
+    _, predictive_covariance = model.predict_joint(all_query_points)
+    expected_covariance = predictive_covariance[0, :8, 8:]
+
+    actual_covariance = model.covariance_between_points(  # type: ignore
+        query_points_1, query_points_2
+    )
+
+    np.testing.assert_allclose(expected_covariance, actual_covariance[0], atol=1e-4)
+
+
+@random_seed
+@pytest.mark.parametrize("whiten", [True, False])
+@pytest.mark.parametrize(
```

Comment:
@uri-granta is this written well enough? (I wrote this test so don't worry you can't upset Henry) 

See also the `two_output_svgp_model` function